### PR TITLE
Model fitting table

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request uses the table subcomponent to log results from model fitting.

[updated model fitting plugin docs](https://jdaviz--2093.org.readthedocs.build/en/2093/specviz/plugins.html#model-fitting)

#### Caveats:
* does not store initial state ~or fixed state~ of parameters (do we want this?)
* ~does not account for the possibility of different model components having the same parameter name and index combination.  This just stores the information as it is stored internally in astropy (see video for example)~
* "model" is stored when it is added.  If the user overwrites an existing model, there will be two table entries with the same model label and would need to realize that the later one corresponds to the one currently loaded in the app (and available via `viz.get_models()`.
* doesn't really translate well to cube fit models (~currently resulting in failed tests~, cube fits are not added to the table)... how do we want to represent these in the table?  ~Or do we just skip them?~  Or transpose the information and have each parameter be an array (single parameter-value per-fitted-pixel)?

https://user-images.githubusercontent.com/877591/225989685-a572d0d6-0e64-439c-9ade-eb9dfe74f488.mov

#### TODO:
- [x] add support for uncertainties (https://github.com/spacetelescope/jdaviz/pull/2093#issuecomment-1474312662)
- [x] handle (or skip) cube fit case
- [x] documentation
- [x] test coverage

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

```
* Model fitting results are logged in a table within the plugin [#2093].
```

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by bmorris3** [comment]
Neat! Do you plan to add columns for uncertainties? (They probably don't need to be displayed by default.)

**Comment by kecnry** [comment]
> Do you plan to add columns for uncertainties? (They probably don't need to be displayed by default.)

That should be pretty easy to do - it would just take a little bit of extra logic to add them when they exist and skip when they don't.  We could just have columns names `parameter_0:uncert` (or similar).  Or we could try to pack the uncertainty into the same column with some sort of uncertainty object, but that might be less convenient for the user when they just want to access the face-value. 🤷‍♂️

**Comment by camipacifici** [comment]
Looks great! Thank you!
A few thoughts:
- It would be nice to see the equation or better the component names somewhere in the table (like what `get_model_parameters()` does).
- If I first create a model with a 3rd order polynomial and then a new model with a 4th order polynomial, `c4` in the first raw is populated with Nan (ok). If after the second model, I create a new model with a 2nd order polynomial, `c3` and `c4` are 0 (technically correct, but different from the behavior in the first row).
- The table could explode if the user calls the same model in combination with different models, like (slope, intercept), (slope_0, intercept_0), (slope_1, intercept_1). I do not have a good solution for this. It is intuitive as it is now, but definitely something to think about.
- Same comment as Brett about uncertainties.
- I do not mind if there are multiple entries for the same model. Hopefully, there will be the possibility to delete rows in the future.
- I have to think about cubes...
- I hit a bug where model fitting tells me a subset is not in the range of the data, but it is. I will check if it is in main.

**Comment by kecnry** [comment]
> It would be nice to see the equation or better the component names somewhere in the table (like what `get_model_parameters()` does).

Storing the equation is easy enough.  Instead of `intercept_0` (which is what astropy uses under-the-hood and so is what ultimately is exposed if the user ever accesses the model object), we could show `L:intercept` (where `L` is the user-provided component label) or  perhaps `L:intercept_0` to keep that connection back to the astropy object.  There are consequences to both of these if the user renames model components though that we should think carefully about.  I also don't see anything like this done elsewhere, although `viz.plugins['Model Fitting'].get_model_component()` does accept the user-friendly label and exposes quite a bit of nice information which we could also include in the table here if we wanted:

```
>>> viz.plugins['Model Fitting'].get_model_component('L')
{'model_type': 'Linear1D',
 'parameters': {'slope': {'value': 2.283977883270026,
   'unit': 'MJy / (sr um)',
   'fixed': False},
  'intercept': {'value': -34122.22032358639,
   'unit': 'MJy / sr',
   'fixed': False}}}
```

How does everyone feel about `L:intercept_0` to store the value and then `L:intercept_0:unit`, `L:intercept_0:fixed`, `L:intercept_0:uncertainty` as other columns which are not shown in the UI by default?  If you rename or re-order columns (so L no longer matches index 0), then future rows will create new columns for those entries (unless you clear the table/row first, of course). 

> If I first create a model with a 3rd order polynomial and then a new model with a 4th order polynomial, c4 in the first raw is populated with Nan (ok). If after the second model, I create a new model with a 2nd order polynomial, c3 and c4 are 0 (technically correct, but different from the behavior in the first row).

I think this behavior will be affected by the decision above, but this is a good use-case to test on whatever we settle on and to make sure we think the result still makes sense.

> The table could explode if the user calls the same model in combination with different models, like (slope, intercept), (slope_0, intercept_0), (slope_1, intercept_1). I do not have a good solution for this. It is intuitive as it is now, but definitely something to think about.
> I do not mind if there are multiple entries for the same model. Hopefully, there will be the possibility to delete rows in the future.

Agreed on both of these - things will get unwieldy (for the user, luckily not the code) when mixing and matching a bunch of different model runs.  But hopefully anyone that intends to export and use the table will make use of keeping it clean for themselves 🤞  Right now you can clear the whole table, but we are planning to support removing individual rows in the future.

> I hit a bug where model fitting tells me a subset is not in the range of the data, but it is. I will check if it is in main.

I vaguely remember a bug like this in mosviz, and even if this is different, would be very surprised if it was introduced here.

**Comment by pllim** [comment]
1. Should we take advantage of ASDF serialization of models for this work? It is all already possible. That is what GWCS uses.
2. Adding arbitrary columns are expensive. I wonder if we should rethink how the table is stored if performance becomes an issue.

**Comment by camipacifici** [comment]
The bug is the same in main. I was just seeing something slightly different than what I remembered, but I double-checked and it is the same.

**Comment by kecnry** [comment]
How about something like this for exposing units/fixed/uncertainties as well as the user-friendly component labels?

https://user-images.githubusercontent.com/877591/226037124-bf10c7b8-d374-4435-8cbe-d2a64f802925.mov

**Comment by kecnry** [comment]
> Should we take advantage of ASDF serialization of models for this work?
> Adding arbitrary columns are expensive. I wonder if we should rethink how the table is stored if performance becomes an issue.

It's not obvious to me how ASDF serialization would help here - we're trying to both show these in the UI in a user-friendly way but also expose in a consistent table-format for exporting to the notebooks, sorting, etc.  All model instances are already stored internally and accessible with `viz.get_models()`, but maybe replacing that with a serialization somewhere could free up some memory, I'm not sure.  It does bug me how we allow access to these things in about 5 different ways... at some point I think we should sit down and deprecate some and try to be more self-consistent.  Is that time now in this PR?   

The table added here in this PR are just some floats and strings, I can't imagine it would become a performance issue before the fitting itself (for a large number of components), but if you can show a case where this is a bottleneck, let me know how to reproduce!

**Comment by pllim** [comment]
If we take advantage of the YAML serialization already provided by ASDF, maybe we can remove a lot of our custom serialization code...

👍 to deprecating redundant APIs.

As for performance, I am not sure either as I am not a power user of this feature. But adding columns for every possible parameter names out there might come back to bite us.

**Comment by kecnry** [comment]
NOTE: I don't expect this to get in for 3.4, so will hold off on adding the change log entry which will only result in the need for a rebase.  See description for proposed change log entry which will be added before merge.

File: docs/specviz/plugins.rst
Comment (bmorris3) [inline docs/specviz/plugins.rst:106]:
```suggestion
Parameter values for each fit are stored in the plugin table.  
```

File: jdaviz/core/template_mixin.py
Comment (bmorris3) [inline jdaviz/core/template_mixin.py:2307]: Is this method effectively setting a default that gets overwritten [elsewhere](https://github.com/spacetelescope/jdaviz/pull/2093/files#diff-f8d8ca5f492ba5b9a6662f90b39d7e11ccff5ee72bce742257a154c13e540744R167)?

**Review by bmorris3 — COMMENTED** [review]
I like it! Some questions below.

**Comment by bmorris3** [comment]
Changelog entry reminder.

File: jdaviz/core/template_mixin.py
Comment (kecnry) [reply jdaviz/core/template_mixin.py:2307]: exactly.  We can override the whole method to a lambda function if we want specific columns to be hidden by default.  If left alone, all added columns will be shown in the UI.

File: docs/specviz/plugins.rst
Comment (kecnry) [reply docs/specviz/plugins.rst:106]: I agree "fitting run" is a bit awkward, but I worry "fit" suggests that they are tied specifically to the output model object, which could be misleading.  Each time "calculate fit" is clicked, a new entry is added.... and if you overwrite the model, you still have a history of the old entry.

**Comment by pllim** [comment]
I got too adventurous and found an unrelated bug at https://github.com/spacetelescope/jdaviz/issues/2108 . I should stick to Specviz (1D) for now...

**Comment by pllim** [comment]
Still not completely convinced that a table is what we really want here. Also, when you overwrite an existing model, the table shows the old one too, but not `specviz.get_models()`, which might cause UX confusion?

Screenshot 2023-03-22 125137

I was also trying to fit models to another spectrum I added hackily and ran into unrelated problem, so I cannot show that in the screenshot. But I wanted to fit yet another model to the second spectrum and then overwrite the `model` again. At that point, I can see how having this table can both help half the users but also maybe confuse the other half. 

Also, you think we should add a timestamp column to the export table showing the time stamp of when each model was generated?

File: docs/specviz/plugins.rst
Comment (bmorris3) [reply docs/specviz/plugins.rst:106]: I know what you mean, but I still think the verb to describe "every time you click 'calculate fit'" should be "fit".

**Review by pllim — APPROVED** [review]
Approving by proxy. Cami says :shipit: . Thanks!

**Pull request merged**
